File: packages/pdp-extractor-agent/src/llm/providers/openai.ts
import { createImageOcrPrompt, createKeywordClassificationPromptParts } from "../prompt";
import { openAiImageOcrTextFormat, openAiKeywordClassificationTextFormat } from "../schemas";
import type {
  ImageTextExtractionRequest,
  ImageTextExtractionResponse,
  KeywordClassificationRequest,
  KeywordClassificationResponse,
  KeywordClassifier,
  LlmProviderConfig
} from "../types";
import type { AiTokenUsage } from "../../types";
import {
  downloadImageAsDataUrl,
  extractJsonObjectText,
  fetchWithTimeout,
  isUnsupportedStructuredOutputError,
  mergeTokenUsages,
  numberField,
  parseImageOcrPayloadText,
  resolveImageInputs,
  responseErrorSuffix,
  temperatureBody
} from "./shared";

const RESPONSES_TIMEOUT_MS = 300_000;

/** OpenAI Responses API adapter for OCR keyword classification. */
export class OpenAIKeywordClassifier implements KeywordClassifier {
  constructor(private readonly config: LlmProviderConfig) {}

  async classifyKeywords(request: KeywordClassificationRequest): Promise<KeywordClassificationResponse> {
    if (!this.config.apiKey) {
      throw new Error("OPENAI_API_KEY is required for the OpenAI keyword classifier.");
    }
    if (!this.config.model) {
      throw new Error("OPENAI_MODEL is required for the OpenAI keyword classifier.");
    }

    const prompt = createKeywordClassificationPromptParts(request);
    const payload = await this.requestResponsesJson({
      model: this.config.model,
      instructions: prompt.system,
      input: prompt.user,
      ...temperatureBody(this.config.temperature),
      text: openAiKeywordClassificationTextFormat
    }, "OpenAI keyword classification");

    return parseProviderJson(payload);
  }

  async extractImageTexts(request: ImageTextExtractionRequest): Promise<ImageTextExtractionResponse> {
    if (!this.config.apiKey) {
      throw new Error("OPENAI_API_KEY is required for OpenAI image OCR.");
    }
    if (!this.config.model) {
      throw new Error("OPENAI_MODEL is required for OpenAI image OCR.");
    }

    const directImages = resolveImageInputs(request);

    try {
      const direct = await this.requestImageOcr(request, directImages);
      const missingInputs = imageInputsWithoutExtractedText(directImages, direct.images);

      if (missingInputs.length === 0) {
        return direct;
      }

      const fallback = await this.extractImageTextsWithDownloadedImages(
        { ...request, imageUrls: missingInputs.map((input) => input.displayUrl), imageInputs: missingInputs },
        new Error(`${missingInputs.length} remote image OCR result(s) returned no readable text.`)
      );

      return fallback.images.length > 0
        ? mergeImageTextExtractionResponses(direct, fallback)
        : direct;
    } catch (directError) {
      if (isQuotaOrBillingError(directError)) {
        throw directError;
      }

      const fallback = await this.extractImageTextsWithDownloadedImages(request, directError);

      if (fallback.images.length > 0) {
        return fallback;
      }

      throw directError;
    }
  }

  private async requestImageOcr(
    request: ImageTextExtractionRequest,
    images: Array<{ displayUrl: string; inputUrl: string }>
  ): Promise<ImageTextExtractionResponse> {
    const content = createImageOcrContent(request, images);
    const payload = await this.requestResponsesJson({
      model: this.config.model,
      input: [
        {
          role: "user",
          content
        }
      ],
      ...temperatureBody(this.config.temperature),
      text: openAiImageOcrTextFormat
    }, "OpenAI image OCR");

    return parseImageOcrJson(payload, images.map((image) => image.displayUrl));
  }

  /**
   * Posts a Responses API body, retrying once without the structured-output
   * `text` format (and once without `temperature`) when the target model
   * rejects those request fields, so older deployments keep working.
   */
  private async requestResponsesJson(body: Record<string, unknown>, label: string): Promise<Record<string, unknown>> {
    let currentBody = body;
    let response = await this.postResponses(currentBody, label);

    if (!response.ok) {
      const suffix = await responseErrorSuffix(response);

      if (currentBody.text !== undefined && isUnsupportedStructuredOutputError(suffix)) {
        const { text: _text, ...retryBody } = currentBody;
        currentBody = retryBody;
        response = await this.postResponses(currentBody, label);
      } else if (currentBody.temperature !== undefined && isUnsupportedTemperatureError(suffix)) {
        const { temperature: _temperature, ...retryBody } = currentBody;
        currentBody = retryBody;
        response = await this.postResponses(currentBody, label);
      } else {
        throw new Error(`${label} failed: ${response.status}${suffix}`);
      }
    }

    if (!response.ok) {
      throw new Error(`${label} failed: ${response.status}${await responseErrorSuffix(response)}`);
    }

    return JSON.parse(await response.text()) as Record<string, unknown>;
  }

  private postResponses(body: Record<string, unknown>, label: string): Promise<Response> {
    return fetchWithTimeout("https://api.openai.com/v1/responses", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.config.apiKey}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify(body)
    }, RESPONSES_TIMEOUT_MS, label);
  }

  private async extractImageTextsWithDownloadedImages(
    request: ImageTextExtractionRequest,
    directError: unknown
  ): Promise<ImageTextExtractionResponse> {
    const images: ImageTextExtractionResponse["images"] = [];
    const errors: string[] = [];
    const usages: AiTokenUsage[] = [];

    for (const input of resolveImageInputs(request)) {
      if (input.inputUrl.startsWith("data:")) {
        try {
          const direct = await this.requestImageOcr(request, [input]);
          images.push(...direct.images);
          if (direct.usage) {
            usages.push(direct.usage);
          }
          if (direct.images.length === 0) {
            errors.push(direct.rawText ? `Inline image OCR returned no readable text: ${direct.rawText.slice(0, 240)}` : "Inline image OCR returned no readable text.");
          }
        } catch (error) {
          if (isQuotaOrBillingError(error)) {
            throw error;
          }
          errors.push(error instanceof Error ? error.message : String(error));
        }
        continue;
      }

      try {
        const dataUrl = await downloadImageAsDataUrl(input.inputUrl);
        const extracted = await this.requestImageOcr(request, [{ displayUrl: input.displayUrl, inputUrl: dataUrl }]);
        images.push(...extracted.images);
        if (extracted.usage) {
          usages.push(extracted.usage);
        }
      } catch (error) {
        if (isQuotaOrBillingError(error)) {
          throw error;
        }
        errors.push(error instanceof Error ? error.message : String(error));
      }
    }

    return {
      images,
      rawText: [
        directError instanceof Error ? directError.message : String(directError),
        ...errors
      ].join("\n"),
      usage: mergeTokenUsages(usages)
    };
  }
}

function responsesOutputText(payload: Record<string, unknown>): string {
  if (typeof payload.output_text === "string") {
    return payload.output_text;
  }
  const output = payload.output as Array<{ content?: Array<{ text?: string }> }> | undefined;
  return output?.flatMap((item) => item.content ?? []).map((item) => item.text ?? "").join("\n") ?? "";
}

function parseProviderJson(payload: Record<string, unknown>): KeywordClassificationResponse {
  const rawText = responsesOutputText(payload);
  const jsonText = extractJsonObjectText(rawText);
  const result = jsonText
    ? JSON.parse(jsonText) as KeywordClassificationResponse
    : { keywords: [], summary: "No parseable JSON returned.", rawText };
  return {
    ...result,
    usage: tokenUsageFromOpenAi(payload.usage)
  };
}

function createImageOcrContent(
  request: ImageTextExtractionRequest,
  images: Array<{ displayUrl: string; inputUrl: string }>
) {
  return [
    {
      type: "input_text",
      text: createImageOcrPrompt({
        ...request,
        imageUrls: images.map((image) => image.displayUrl)
      })
    },
    ...images.flatMap((image, index) => [
      {
        type: "input_text",
        text: `Image ${index + 1}: ${image.displayUrl}`
      },
      {
        type: "input_image",
        image_url: image.inputUrl,
        detail: "high"
      }
    ])
  ];
}

function parseImageOcrJson(payload: Record<string, unknown>, imageUrls: string[]): ImageTextExtractionResponse {
  return {
    ...parseImageOcrPayloadText(responsesOutputText(payload), imageUrls),
    usage: tokenUsageFromOpenAi(payload.usage)
  };
}

function imageInputsWithoutExtractedText(
  requestedInputs: Array<{ displayUrl: string; inputUrl: string }>,
  extractedImages: ImageTextExtractionResponse["images"]
): Array<{ displayUrl: string; inputUrl: string }> {
  const extractedUrls = new Set(extractedImages.map((image) => image.imageUrl));
  return requestedInputs.filter((input) => !extractedUrls.has(input.displayUrl));
}

function mergeImageTextExtractionResponses(
  primary: ImageTextExtractionResponse,
  fallback: ImageTextExtractionResponse
): ImageTextExtractionResponse {
  const seen = new Set<string>();
  const images = [...primary.images, ...fallback.images].filter((image) => {
    const key = image.imageUrl;
    if (seen.has(key)) {
      return false;
    }
    seen.add(key);
    return image.text.trim().length > 0;
  });

  return {
    images,
    rawText: [primary.rawText, fallback.rawText].filter(Boolean).join("\n"),
    usage: mergeTokenUsages([primary.usage, fallback.usage].filter((usage): usage is AiTokenUsage => Boolean(usage)))
  };
}

function tokenUsageFromOpenAi(value: unknown): AiTokenUsage | undefined {
  if (!value || typeof value !== "object") {
    return undefined;
  }
  const usage = value as Record<string, unknown>;
  const inputTokens = numberField(usage.input_tokens) ?? numberField(usage.prompt_tokens);
  const outputTokens = numberField(usage.output_tokens) ?? numberField(usage.completion_tokens);
  const totalTokens = numberField(usage.total_tokens);
  return inputTokens !== undefined || outputTokens !== undefined || totalTokens !== undefined
    ? { inputTokens, outputTokens, totalTokens }
    : undefined;
}

function isUnsupportedTemperatureError(message: string): boolean {
  return /unsupported value[^]*temperature|temperature[^]*(?:unsupported|only the default)/i.test(message);
}

function isQuotaOrBillingError(error: unknown): boolean {
  const message = error instanceof Error ? error.message : String(error);
  return /(exceeded your current quota|insufficient_quota|billing|check your plan|rate limit|too many requests)/i.test(message);
}

File: packages/pdp-extractor-agent/src/pdp_extractor_agent/providers/openai.py
"""OpenAI Responses API provider compatibility adapter."""

from __future__ import annotations

import base64
import json
import math
import re
from collections.abc import Awaitable, Callable, Mapping
from typing import Any, cast

import httpx
from neo_js_compat import js_json_bytes

from .._json_types import as_list, as_mapping
from ..ocr.layout import parse_image_ocr_payload_text
from ..ocr.prompt import create_image_ocr_prompt, create_keyword_classification_prompt_parts
from ..schemas import OPENAI_IMAGE_OCR_TEXT_FORMAT, OPENAI_KEYWORD_CLASSIFICATION_TEXT_FORMAT
from .transport import (
    DEFAULT_MODEL_TIMEOUT_SECONDS,
    is_unsupported_structured_output_error,
    model_timeout_seconds,
    resolve_image_inputs,
    response_json_object,
    temperature_body,
    timeout_label_seconds,
)

ImageFetcher = Callable[[str], Awaitable[tuple[str, bytes]]]


class OpenAIProvider:
    """Responses API adapter for OCR keyword classification and vision OCR."""

    def __init__(
        self,
        *,
        api_key: object,
        model: object,
        temperature: object = None,
        transport: object = None,
        image_fetcher: object = None,
        timeout_seconds: object = DEFAULT_MODEL_TIMEOUT_SECONDS,
    ) -> None:
        self.api_key = api_key
        self.model = model
        self.temperature = temperature
        self.transport = transport
        self.image_fetcher = image_fetcher
        self.timeout_seconds = model_timeout_seconds(timeout_seconds)

    async def classify_keywords(self, request: Mapping[str, Any]) -> dict[str, Any]:
        self._require("keyword classifier")
        prompt = create_keyword_classification_prompt_parts(request)
        payload = await self._request_responses_json(
            {
                "model": _text(self.model),
                "instructions": prompt["system"],
                "input": prompt["user"],
                **temperature_body(self.temperature),
                "text": OPENAI_KEYWORD_CLASSIFICATION_TEXT_FORMAT,
            },
            "OpenAI keyword classification",
        )
        result = _parse_keyword_payload(_responses_output_text(payload))
        usage = _token_usage(as_mapping(payload.get("usage")))
        if usage is not None:
            result["usage"] = usage
        return result

    async def classifyKeywords(self, request: Mapping[str, Any]) -> dict[str, Any]:
        return await self.classify_keywords(request)

    async def extract_image_text(self, request: Mapping[str, Any]) -> dict[str, Any]:
        self._require("image OCR")
        inputs = resolve_image_inputs(request)
        try:
            direct = await self._request_image_ocr(request, inputs)
            missing = [item for item in inputs if item["displayUrl"] not in _extracted_urls(direct)]
            if not missing:
                return direct
            fallback = await self._downloaded_fallback(
                {**dict(request), "imageUrls": [item["displayUrl"] for item in missing], "imageInputs": missing},
                RuntimeError(f"{len(missing)} remote image OCR result(s) returned no readable text."),
            )
            return _merge_image_responses(direct, fallback) if _has_image_text(fallback) else direct
        except Exception as direct_error:
            if _is_quota_or_billing_error(str(direct_error)):
                raise
            fallback = await self._downloaded_fallback(request, direct_error)
            if _has_image_text(fallback):
                return fallback
            raise

    async def extractImageTexts(self, request: Mapping[str, Any]) -> dict[str, Any]:
        return await self.extract_image_text(request)

    async def _request_image_ocr(self, request: Mapping[str, Any], inputs: list[dict[str, str]]) -> dict[str, Any]:
        content: list[dict[str, Any]] = [
            {
                "type": "input_text",
                "text": create_image_ocr_prompt({**dict(request), "imageUrls": [item["displayUrl"] for item in inputs]}),
            }
        ]
        for index, item in enumerate(inputs, start=1):
            content.extend(
                [
                    {"type": "input_text", "text": f"Image {index}: {item['displayUrl']}"},
                    {"type": "input_image", "image_url": item["inputUrl"], "detail": "high"},
                ]
            )
        payload = await self._request_responses_json(
            {
                "model": _text(self.model),
                "input": [{"role": "user", "content": content}],
                **temperature_body(self.temperature),
                "text": OPENAI_IMAGE_OCR_TEXT_FORMAT,
            },
            "OpenAI image OCR",
        )
        result = parse_image_ocr_payload_text(_responses_output_text(payload), [item["displayUrl"] for item in inputs])
        usage = _token_usage(as_mapping(payload.get("usage")))
        if usage is not None:
            result["usage"] = usage
        return result

    async def _downloaded_fallback(self, request: Mapping[str, Any], direct_error: BaseException) -> dict[str, Any]:
        images: list[dict[str, Any]] = []
        errors: list[str] = []
        usages: list[Mapping[str, Any]] = []
        for input_ in resolve_image_inputs(request):
            try:
                data_input = input_["inputUrl"] if input_["inputUrl"].startswith("data:") else await self._download_image_as_data_url(input_["inputUrl"])
                extracted = await self._request_image_ocr(
                    request, [{"displayUrl": input_["displayUrl"], "inputUrl": data_input}]
                )
                images.extend(_image_rows(extracted))
                usage = as_mapping(extracted.get("usage"))
                if usage is not None:
                    usages.append(usage)
                if not _has_image_text(extracted):
                    raw_text = extracted.get("rawText")
                    errors.append(
                        f"Inline image OCR returned no readable text: {str(raw_text)[:240]}"
                        if isinstance(raw_text, str) and raw_text
                        else "Inline image OCR returned no readable text."
                    )
            except Exception as error:
                if _is_quota_or_billing_error(str(error)):
                    raise
                errors.append(str(error))
        result: dict[str, Any] = {"images": images, "rawText": "\n".join([str(direct_error), *errors])}
        usage = _merge_usages(usages)
        if usage is not None:
            result["usage"] = usage
        return result

    async def _request_responses_json(self, body: Mapping[str, Any], label: str) -> Mapping[str, Any]:
        current = dict(body)
        response = await self._post(current, label)
        if not response.is_success:
            suffix = _response_error_suffix(response)
            if "text" in current and is_unsupported_structured_output_error(suffix):
                current.pop("text", None)
                response = await self._post(current, label)
            elif "temperature" in current and _unsupported_temperature(suffix):
                current.pop("temperature", None)
                response = await self._post(current, label)
            else:
                raise RuntimeError(f"{label} failed: {response.status_code}{suffix}")
        if not response.is_success:
            suffix = _response_error_suffix(response)
            raise RuntimeError(f"{label} failed: {response.status_code}{suffix}")
        return response_json_object(response, label)

    async def _post(self, body: Mapping[str, Any], label: str) -> httpx.Response:
        try:
            async with httpx.AsyncClient(
                transport=cast(httpx.AsyncBaseTransport | None, self.transport),
                timeout=self.timeout_seconds,
                follow_redirects=True,
            ) as client:
                return await client.post(
                    "https://api.openai.com/v1/responses",
                    headers={"Authorization": f"Bearer {_text(self.api_key)}", "Content-Type": "application/json"},
                    content=js_json_bytes(body),
                )
        except httpx.TimeoutException as error:
            raise RuntimeError(f"{label} timed out after {timeout_label_seconds(self.timeout_seconds)}s.") from error

    async def _download_image_as_data_url(self, image_url: str) -> str:
        if callable(self.image_fetcher):
            mime_type, content = await cast(ImageFetcher, self.image_fetcher)(image_url)
            return _data_url(mime_type, content)
        try:
            async with httpx.AsyncClient(
                transport=cast(httpx.AsyncBaseTransport | None, self.transport), timeout=60.0, follow_redirects=True
            ) as client:
                response = await client.get(
                    image_url,
                    headers={
                        "Accept": "image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8",
                        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0 Safari/537.36",
                    },
                )
        except httpx.TimeoutException as error:
            raise RuntimeError("Image download for OCR timed out after 60s.") from error
        if not response.is_success:
            raise RuntimeError(f"Image download failed: {response.status_code}{_response_error_suffix(response)}")
        mime_type = response.headers.get("content-type", "image/jpeg").split(";", 1)[0].strip() or "image/jpeg"
        if not mime_type.startswith("image/"):
            raise RuntimeError(f"Image download returned non-image content-type: {mime_type}")
        if len(response.content) > 10 * 1024 * 1024:
            raise RuntimeError(f"Image is too large for OCR fallback: {len(response.content)} bytes")
        return _data_url(mime_type, response.content)

    def _require(self, label: str) -> None:
        if not self.api_key:
            raise ValueError(
                "OPENAI_API_KEY is required for OpenAI image OCR."
                if label == "image OCR"
                else "OPENAI_API_KEY is required for the OpenAI keyword classifier."
            )
        if not self.model:
            raise ValueError(
                "OPENAI_MODEL is required for OpenAI image OCR."
                if label == "image OCR"
                else "OPENAI_MODEL is required for the OpenAI keyword classifier."
            )


def _responses_output_text(payload: Mapping[str, Any]) -> str:
    output_text = payload.get("output_text")
    if isinstance(output_text, str):
        return output_text
    parts: list[str] = []
    for raw in as_list(payload.get("output")) or []:
        item = as_mapping(raw)
        contents = as_list(item.get("content")) if item is not None else None
        for raw_content in contents or []:
            content = as_mapping(raw_content)
            text = content.get("text") if content is not None else None
            if isinstance(text, str):
                parts.append(text)
    return "\n".join(parts)


def _parse_keyword_payload(raw_text: str) -> dict[str, Any]:
    json_text = _extract_json_object_text(raw_text)
    if json_text is None:
        return {"keywords": [], "summary": "No parseable JSON returned.", "rawText": raw_text}
    # A braced candidate is an attempted structured response.  The retained
    # adapter lets JSON.parse fail here; only prose/no-object output receives
    # the tolerant empty envelope above.
    return dict(as_mapping(json.loads(json_text)) or {})


def _extract_json_object_text(raw_text: str) -> str | None:
    text = raw_text.strip()
    if text.startswith("{") and text.endswith("}"):
        return text
    fenced = re.search(r"```(?:json)?\s*([\s\S]*?)\s*```", text, re.IGNORECASE)
    if fenced is not None and fenced.group(1).strip().startswith("{"):
        return fenced.group(1).strip()
    match = re.search(r"\{[\s\S]*\}", text)
    return match.group(0) if match is not None else None


def _token_usage(value: Mapping[str, Any] | None) -> dict[str, int | float] | None:
    if value is None:
        return None
    input_tokens = _number_field(value.get("input_tokens"))
    if input_tokens is None:
        input_tokens = _number_field(value.get("prompt_tokens"))
    output_tokens = _number_field(value.get("output_tokens"))
    if output_tokens is None:
        output_tokens = _number_field(value.get("completion_tokens"))
    total_tokens = _number_field(value.get("total_tokens"))
    result: dict[str, int | float] = {}
    if input_tokens is not None:
        result["inputTokens"] = input_tokens
    if output_tokens is not None:
        result["outputTokens"] = output_tokens
    if total_tokens is not None:
        result["totalTokens"] = total_tokens
    return result or None


def _number_field(value: object) -> int | float | None:
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value):
        return None
    return value


def _image_rows(result: Mapping[str, Any]) -> list[dict[str, Any]]:
    return [dict(item) for raw in as_list(result.get("images")) or [] if (item := as_mapping(raw)) is not None]


def _extracted_urls(result: Mapping[str, Any]) -> set[str]:
    return {
        str(item.get("imageUrl"))
        for item in _image_rows(result)
        if isinstance(item.get("text"), str) and item["text"].strip()
    }


def _has_image_text(result: Mapping[str, Any]) -> bool:
    return any(isinstance(item.get("text"), str) and item["text"].strip() for item in _image_rows(result))


def _merge_image_responses(primary: Mapping[str, Any], fallback: Mapping[str, Any]) -> dict[str, Any]:
    seen: set[str] = set()
    images: list[dict[str, Any]] = []
    for item in [*_image_rows(primary), *_image_rows(fallback)]:
        image_url, text = item.get("imageUrl"), item.get("text")
        if isinstance(image_url, str) and isinstance(text, str) and text.strip() and image_url not in seen:
            seen.add(image_url)
            images.append(item)
    result: dict[str, Any] = {
        "images": images,
        "rawText": "\n".join(
            value for value in (primary.get("rawText"), fallback.get("rawText")) if isinstance(value, str) and value
        ),
    }
    usage = _merge_usages([usage for candidate in (primary, fallback) if (usage := as_mapping(candidate.get("usage"))) is not None])
    if usage is not None:
        result["usage"] = usage
    return result


def _merge_usages(usages: list[Mapping[str, Any]]) -> dict[str, int | float] | None:
    totals: dict[str, int | float] = {}
    for usage in usages:
        for key in ("inputTokens", "outputTokens", "totalTokens"):
            number = _number_field(usage.get(key))
            if number is not None:
                totals[key] = totals.get(key, 0) + number
    return totals or None


def _data_url(mime_type: str, content: bytes) -> str:
    return f"data:{mime_type};base64,{base64.b64encode(content).decode('ascii')}"


def _response_error_suffix(response: httpx.Response) -> str:
    cleaned = " ".join(response.text.split())
    return f" - {cleaned[:500]}" if cleaned else ""


def _unsupported_temperature(message: str) -> bool:
    return re.search(r"unsupported value[\s\S]*temperature|temperature[\s\S]*(?:unsupported|only the default)", message, re.IGNORECASE) is not None


def _is_quota_or_billing_error(message: str) -> bool:
    return re.search(
        r"exceeded your current quota|insufficient_quota|billing|check your plan|rate limit|too many requests",
        message,
        re.IGNORECASE,
    ) is not None


def _text(value: object) -> str:
    return value if isinstance(value, str) else str(value or "")


OpenAIKeywordClassifier = OpenAIProvider
